File: shims/qpid-proton-python/src/amqp_features_test/Sender.py
#!/usr/bin/env python

"""
AMQP features test sender shim for qpid-interop-test
"""

#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

from json import dumps, loads
import os.path
import sys
from traceback import format_exc

from proton.handlers import MessagingHandler
from proton.reactor import Container

class AmqpFeaturesTestSender(MessagingHandler):
    """
    Sender shim for AMQP dtx test
    ...
    """
    def __init__(self, broker_url, queue_name, test_type, test_args):
        super(AmqpFeaturesTestSender, self).__init__()
        self.broker_url = broker_url
        self.queue_name = queue_name
        self.test_type = test_type
        self.test_args = test_args
        self.test_value_list = []
        self.sent = 0
        self.confirmed = 0
        self.total = len(self.test_value_list)
        self.remote_properties = None

    def on_start(self, event):
        """Event callback for when the client starts"""
        connection = event.container.connect(url=self.broker_url, sasl_enabled=False)
        event.container.create_sender(connection, target=self.queue_name)

    def on_connection_remote_open(self, event):
        """Callback for remote connection open"""
        self.remote_properties = event.connection.remote_properties
        if self.test_type == 'connection_property':
            event.connection.close()

    def on_sendable(self, event):
        """Event callback for when send credit is received, allowing the sending of messages"""
        if self.sent == 0:
            for test_value in self.test_value_list:
                if event.sender.credit:
                    message = self.create_message(test_value)
                    if message is not None:
                        event.sender.send(message)
                        self.sent += 1
                    else:
                        event.connection.close()
                        return

    def create_message(self, test_value):
        """
        Creates a single message with the test value translated from its string representation to the appropriate
        AMQP value (set in self.amqp_type).
        """
        return None

    def on_accepted(self, event):
        """Event callback for when a sent message is accepted by the broker"""
        self.confirmed += 1
        if self.confirmed == self.total:
            event.connection.close()

    def on_disconnected(self, event):
        """Event callback for when the broker disconnects with the client"""
        self.sent = self.confirmed

    def get_remote_properties(self):
        """Return the remote (broker) properties"""
        return self.remote_properties


# --- main ---
# Args: 1: Broker address (ip-addr:port)
#       2: Queue name
#       3: Test type
#       4: Test args
try:
    SENDER = AmqpFeaturesTestSender(sys.argv[1], sys.argv[2], sys.argv[3], loads(sys.argv[4]))
    Container(SENDER).run()
    print sys.argv[3]
    if sys.argv[3] == 'connection_property':
        print dumps(SENDER.get_remote_properties())
except KeyboardInterrupt:
    pass
except Exception as exc:
    print os.path.basename(sys.argv[0]), 'EXCEPTION:', exc
    print format_exc()
        

File: shims/qpid-proton-cpp/src/qpidit/amqp_features_test/Sender.cpp
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

#include "qpidit/amqp_features_test/Sender.hpp"

#include <iostream>
#include <json/json.h>
#include <proton/container.hpp>
#include <proton/default_container.hpp>
#include <proton/error_condition.hpp>
#include <proton/thread_safe.hpp>
#include <proton/tracker.hpp>
#include <proton/transport.hpp>
#include <qpidit/QpidItErrors.hpp>
#include <stdlib.h> // exit()


namespace qpidit
{
    namespace amqp_features_test
    {
        Sender::Sender(const std::string& brokerUrl,
                       const std::string& queueName,
                       const std::string& testType,
                       const Json::Value& testValues) :
                       _brokerUrl(brokerUrl),
                       _queueName(queueName),
                       _testType(testType),
                       _testValues(testValues)
        {}

        Sender::~Sender() {}

        void Sender::on_container_start(proton::container &c) {
            std::ostringstream oss;
            oss << _brokerUrl << "/" << _queueName;
            c.open_sender(oss.str());
        }

        void Sender::on_connection_open(proton::connection& c) {
            if (_testType.compare("connection_property") == 0) {
                // Python: self.remote_properties = event.connection.remote_properties
            }
        }

        void Sender::on_sendable(proton::sender &s) {
            if (_totalMsgs == 0) {
                s.connection().close();
            } else  {
                if (_testType.compare("connection_property") == 0) {
                    // Do nothing
                } else {
                    // Unknown test
                }
            }
        }

        void Sender::on_tracker_accept(proton::tracker &t) {
            _msgsConfirmed++;
            if (_msgsConfirmed == _totalMsgs) {
                t.connection().close();
            }
        }

        void Sender::on_transport_close(proton::transport &t) {
            _msgsSent = _msgsConfirmed;
        }

        void Sender::on_connection_error(proton::connection &c) {
            std::cerr << "AmqpSender::on_connection_error(): " << c.error() << std::endl;
        }

        void Sender::on_sender_error(proton::sender &s) {
            std::cerr << "AmqpSender::on_sender_error(): " << s.error() << std::endl;
        }

        void Sender::on_session_error(proton::session &s) {
            std::cerr << "AmqpSender::on_session_error(): " << s.error() << std::endl;
        }

        void Sender::on_transport_error(proton::transport &t) {
            std::cerr << "AmqpSender::on_transport_error(): " << t.error() << std::endl;
        }

        void Sender::on_error(const proton::error_condition &ec) {
            std::cerr << "AmqpSender::on_error(): " << ec << std::endl;
        }
    } /* namespace amqp_features_test */
} /* namespace qpidit */


/*
 * --- main ---
 * Args: 1: Broker address (ip-addr:port)
 *       2: Queue name
 *       3: Test type
 *       4: JSON test values
 */

int main(int argc, char** argv) {
    // TODO: improve arg management a little...
    if (argc != 5) {
        throw qpidit::ArgumentError("Incorrect number of arguments");
    }

    try {
        Json::Value testValues;
        Json::Reader jsonReader;
        if (not jsonReader.parse(argv[4], testValues, false)) {
            throw qpidit::JsonParserError(jsonReader);
        }

        qpidit::amqp_features_test::Sender sender(argv[1], argv[2], argv[3], testValues);
        proton::default_container(sender).run();
    } catch (const std::exception& e) {
        std::cerr << "amqp_features_test Sender error: " << e.what() << std::endl;
        exit(1);
    }
    exit(0);
}
